Fragmentation counters, for IPv4 and IPv6

kbara opened the pull request:
This also includes a facility to regenerate test counter files.

Commit:
Fragmentation counters, for IPv4 and IPv6

This also includes a facility to regenerate test counter files.

dpino commented on src/apps/lwaftr/fragmentv4_hardened.lua:
```diff
@@ -267,20 +267,23 @@ function initialize_frag_table(max_fragmented_packets, max_pkt_frag)
    if not ptr then
       local reassembly_buf = packet_to_reassembly_buffer(fragment)
-      frags_table:add_with_random_ejection(key, reassembly_buf, false)
+      _, ej = frags_table:add_with_random_ejection(key, reassembly_buf, false)
```
Or can use `select` to pick one of the returned values, for instance:

```
ej = select(2, frags_table:add_with_random_ejection(key, reassembly_buf, false))
```

kbara replied:
I could, but that strikes me as a lot less readable. Call me an Erlanger if you must. :-)

teknico replied:
Python and Go use the same convention, it's not just Erlang. :-)

dpino commented on src/program/lwaftr/check/check.lua:
```diff
@@ -75,6 +79,18 @@ function validate_diff(actual, expected)
 
+local function regen_counters(counters, outfile)
+   local cnames = {}
+   for k in pairs(counters) do table.insert(cnames, k) end
```
lwutil has a keys function.

``` lua
local cnames = lwutil.keys(counters)
```

dpino commented:
LGTM. Maybe is worth to wait for a second review from @teknico

Commit:
Use lwutil.keys

kbara commented:
Definitely. Counter documentation changes are blocked on him sorting out the merge conflicts that have been committed to that file (by him, according to ```git blame```.)

Commit:
Move the line-oriented counters towards the edges of the app network

This allows more accurate reporting in cases where apps process the packets
before the lwaftr can (and after leaving it), including:
a) Fragmentation/reassembly
b) Control-plane messages, including ARP/NDP/pinging the lwaftr itself

kbara commented:
I've moved the 'line oriented' input/output/drop counters into the fragmentation and reassembly apps. The benefit is that they now more closely reflect what goes over the line. There are a number of tradeoffs involved. (To be entirely accurately line-oriented, the counters would have to be in the driver itself - for instance, we never count vlan tags, because they're already stripped - but I'd rather keep the counters card-agnostic.)

Conceptually, this belongs in a separate app, rather than the fragmentation and reassembly apps, but I don't want to add 4 extra apps to the app network. I've placed them where they are because it's at the outer fringes of our app network (excluding the card, and the software vlan tag handlers that we use for testing.)

teknico commented:
Oops, sorry, I wonder how that happened. Making a PR to remove the merge conflicts to `README.counters.md`.

Commit:
Move ctable changes to a wrapper

This makes the only significant change to ctable an extra instance variable
that stores how much memory is being used, and a getter for it.

Also, make random_eject closer to uniform.
As discussed on https://github.com/Igalia/snabb/pull/388/

teknico commented on src/program/lwaftr/check/check.lua:
```diff
@@ -8,8 +8,11 @@ local setup = require("program.lwaftr.setup")
+local write_to_file = lwutil.write_to_file
 
 function show_usage(code)
    print(require("program.lwaftr.check.README_inc"))
```
The README file needs a description of the new `-r` option, and of the existing `on-a-stick` option too.

teknico commented:
Will LGTM once the README is updated.

I wonder how to redraw the counter diagrams taking into account the counters' new place in the code.

Commit:
Document 'check -r'
